File: ESPictureBrowserDemo/ESPictureBrowserDemo/ViewController.m
//
//  ViewController.m
//  ESPictureBrowserDemo
//
//  Created by EnjoySR on 2016/10/16.
//  Copyright © 2016年 EnjoySR. All rights reserved.
//

#import "ViewController.h"
#import "ESPictureModel.h"
#import "ESCellNode.h"
#import <AsyncDisplayKit.h>
#import <PINCache/PINCache.h>
#import <YYWebImage/YYWebImage.h>

@interface ViewController ()<ASTableDelegate, ASTableDataSource>

@property (nonatomic, strong) NSArray *datas;

@property (nonatomic, strong) ASTableView *tableView;
    
@end

@implementation ViewController
    
- (NSArray *)datas {
    if (_datas == nil) {
        NSURL *url = [[NSBundle mainBundle] URLForResource:@"list.json" withExtension:nil];
        NSData *data = [NSData dataWithContentsOfURL:url];
        NSArray *array = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        
        NSMutableArray *result = [NSMutableArray array];
        for (NSArray *value in array) {
            NSMutableArray *arrayM = [NSMutableArray array];
            for (NSDictionary *dict in value) {
                ESPictureModel *model = [ESPictureModel new];
                [model setValuesForKeysWithDictionary:dict];
                [arrayM addObject:model];
            }
            [result addObject:arrayM];
        }
        _datas = [result copy];
    }
    return _datas;
}

- (void)loadView {
    ASTableView *tableView = [ASTableView new];
    tableView.asyncDelegate = self;
    tableView.asyncDataSource = self;
    self.view = tableView;
    self.tableView = tableView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemTrash target:self action:@selector(clearCache)];
}

- (void)clearCache {
    [[YYWebImageManager sharedManager].cache.diskCache removeAllObjects];
    [[YYWebImageManager sharedManager].cache.memoryCache removeAllObjects];
}
    
- (void)dealloc {
    NSLog(@"销毁");
}

#pragma mark - "代理数据源方法"

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.datas.count;
}

- (ASCellNode *)tableView:(ASTableView *)tableView nodeForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    ESCellNode *node = [ESCellNode new];
    NSArray *pictureModels = self.datas[indexPath.row];
    [node setPictureModels:pictureModels];
    return node;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:true];
}

@end

File: ESPictureBrowserDemo/ESPictureBrowserDemo/ViewController.swift
//
//  ViewController.swift
//  ESPictureBrowserDemo
//
//  Created by EnjoySR on 2016/10/15.
//  Copyright © 2016年 EnjoySR. All rights reserved.
//

import UIKit
import AsyncDisplayKit

class ViewController: UIViewController {
    
    override func loadView() {
        view = tableView
    }

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    /// 数据
    lazy var datas: [[ESPictureModel]] = {
        // 获取数据
        let url = Bundle.main.url(forResource: "list.json", withExtension: nil)!
        let data = try! Data(contentsOf: url)
        let array = try! JSONSerialization.jsonObject(with: data, options: []) as! [[[String: Any]]]
        
        var result = [[ESPictureModel]]()
        // 遍历字典转模型
        for value in array {
            
            var arrayM = [ESPictureModel]()
            for dict in value {
                let model = ESPictureModel()
                model.setValuesForKeys(dict)
                arrayM.append(model)
            }
            result.append(arrayM)
        }
        return result
    }()
    
    lazy var tableView: ASTableView = {
        let tableView = ASTableView()
        tableView.asyncDelegate = self
        tableView.asyncDataSource = self
        return tableView
    }()
}

extension ViewController: ASTableViewDataSource, ASTableViewDelegate {
    
    func tableView(_ tableView: ASTableView, nodeForRowAt indexPath: IndexPath) -> ASCellNode {
        let cellNode = ESCellNode()
        let model = datas[indexPath.row]
        cellNode.configurModel(pictureModels: model)
        return cellNode
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return datas.count;
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
}


